set-free-balance.md

```diff
@@ -0,0 +1,150 @@
+/// the caller will do this.
+fn set_free_balance(who: &T::AccountId, balance: T::Balance) -> UpdateBalanceOutcome {
+  // Commented out for now - but consider it instructive.
+  // assert!(!Self::total_balance(who).is_zero());
+  // assert!(Self::free_balance(who) > T::ExistentialDeposit::get());
+  if balance < T::ExistentialDeposit::get() {
+    <FreeBalance<T, I>>::insert(who, balance);
+    Self::on_free_too_low(who);
+    UpdateBalanceOutcome::AccountKilled
+  } else {
+    <FreeBalance<T, I>>::insert(who, balance);
+    UpdateBalanceOutcome::Updated
+  }
+}
+```
+
+State Model
+-----------
+
+```k
+module SET-FREE-BALANCE-SPEC
+    imports INT
+    imports LIST
+
+    configuration
+      <set-free-balance>
+        <k> $ACTION:Action </k>
+        <events> .List </events>
+        <existentialDeposit> 0 </existentialDeposit>
+        <accounts>
+          <account multiplicity="*" type="Map">
+            <accountID> .AccountId:AccountId </accountID>
+            <balance> 0 | 0 </balance>
+            <nonce> .Nonce </nonce>
+          </account>
+        </accounts>
+      </set-free-balance>
+```
+
+Data
+----
+
+-   An `AccountId` is an optional `Int`.
+-   A `Balance` is a tuple of `Int` (for free/reserved balance).
+-   A `Nonce` is an optional `Int`.
+-   An `Event` records some happenning.
+
+```k
+    syntax AccountId ::= ".AccountId" | Int
+ // ---------------------------------------
+
+    syntax Balance ::= Int "|" Int
+ // ------------------------------
+
+    syntax Nonce ::= ".Nonce" | Int
+ // -------------------------------
+
+    syntax Event ::= DustEvent ( Int )
+ // ----------------------------------
+```
+
+Results
+-------
+
+A `Result` is the return value of an execution step.
+
+-   `AccountKilled` indicates that the free balance goes below the existential threshold.
+-   `Updated` indicates that an account was updated successfully.
+
+```k
+    syntax Result ::= "AccountKilled" | "Updated"
+ // ---------------------------------------------
+```
+
+Actions and Results
+-------------------
+
+An `Action` is an execution step (or the result of an execution step).
+A `Result` is considered an `Action`.
+
+```k
+    syntax Action ::= Result
+ // ------------------------
+```
+
+### `set_free_balance`
+
+-   Updates an accounts balance if the new balance is above the existential threshold.
+-   Kills the account if the balance goes below the existential threshold and the reserved balance is non-zero.
+-   Reaps the account if the balance goes below the existential threshold and the reserved balance is zero.
+
+```k
+    syntax Action ::= set_free_balance ( AccountId , Int )
+ // ------------------------------------------------------
+    rule [account-updated]:
+         <k> set_free_balance(WHO, BALANCE) => Updated </k>
+         <existentialDeposit> EXISTENTIAL_DEPOSIT </existentialDeposit>
+         <account>
+           <accountID> WHO </accountID>
+           <balance> (_ => 0) | _ </balance>
```

Comment:
Looking at the rust code, it seems this should be `BALANCE` rather than `0`

Reply:
Fixed.